design/transaction-layer-refactoring.md

```diff
@@ -0,0 +1,151 @@
+
+If more and more commands are not sharing the same procedure and dependencies, the current structure will have no benefit. Instead, the drawbacks become clearer.
+
+The code for a certain command is spread in too many files. Each function handles a part of work of various commands. People are hard to understand what happens about a single command.
+
+And now, different commands have different dependencies. But because all commands share the same procedure, we must pass in the union of all dependencies of all commands. It makes adding commands and dependencies more difficult.
+
+## New structure
+
+Based on the analysis above, the current structure is not flexible enough and should be dropped. Instead, we can change to put logic of a single command together. Common steps can be extracted as methods for reuse. Then, it will be much easier to find out the procedure of each command while still not repeating code.
+
+Steps like acquiring latches and waiting for raft replication cannot finish immediately, so it is appropriate to make every command an `async fn`.
+
+For example, the `AcquirePessimisticLock` command can be like:
+
+```rust
+async fn acquire_pessimistic_lock(&self, req: PessimisticLockRequest) -> PessimisticLockResponse {
+    let mut resp = PessimisticLockResponse::default();
+    let guards = self.lock_keys(req.mutations.iter().map(|m| m.get_key())).await;
+    let snapshot = self.get_snapshot(req.get_context()).await;
+    
+    let mut txn = MvccTxn::new(...);
+    let mut locked = None;
+    for (mutation, guard) in req.mutations.into_iter().zip(&guards) {
+        match txn.acquire_pessimistic_lock(...) {
+            Ok(...) => ...,
+            Err(KeyIsLocked(lock_info)) => {
+                guard.set_lock(lock_info.clone());
+                let lock_released = guard.lock_released(); // returns a future which is ready when the lock is released
+                locked = Some(lock_info, );
+            },
+            Err(e) => ...
+        }
+    }
+    
+    if let Some((lock_info, lock_released)) = locked {
+        drop(guards); // release the latches first
+        lock_released.await; // easy to add timeout
+        resp.set_errors(...);
+    } else {
+        if self.cfg.pipelined_pessimistic_lock {
+            // write to the raft store asynchronously
+            let engine = self.engine.clone();
+            self.spawn(async move {
+                engine.write(txn.into_modifies()).await;
+                drop(guards);
+            });
+        } else {
+            // write to the raft store synchronously
+            self.engine.write(txn.into_modifies()).await; 
+        }
+        ...
+    }
+    resp
+}
+```
+
+The goal is to put the whole process of each command inside a single function. Then, people only need to look at one function to learn the process. Code related to transactions will be more understandable.

```

Comment:
I think the problem with this approach is that you end up with some very long, complex functions which are difficult to decompose and it is difficult to share code between commands because abstracting over functions is not well-served by Rust. I think having one object per command, which can include multiple async methods has the benefits of this approach, without the drawbacks.

Reply:
I think actually different commands share little code: only acquiring the latch and getting the snapshot.

The form in which code organizes is negotiable. Command structs with its own methods also sound good to me. However, I don't fully understand "abstracting over functions is not well-served by Rust". What's the difficulty of extracting common code into functions?

Reply:
I just mean that it can be a bit less ergonomic - with objects it is easy to encapsulate the types of intermediate values as part of the object, whereas if you pass them to a function, they either must have the same concrete type or you end creating a bunch of tiny, special purpose traits or using some other verbose work around